### PR TITLE
INT-2298 - Add getStepStartStates, ingest additional resources and update docs

### PR DESCRIPTION
**Pull request by eXtremeX**
Changes: 

- [x] Add basic `getStepStartStates`
    - [x] Entitlement set
- [x] Ingest additional resources
    - [x] `tfe_workspace`
    - [x] `tfe_workspace_resource`
    - [x] `tfe_entitlement_set`
    - [x] `tfe_team`
    - [x] `tfe_account`
- [x] Update `docs/`

Notes:
1) We're assuming the idea is for a customer to input organization token (not user/team's one - [all types described here](https://www.terraform.io/cloud-docs/users-teams-organizations/api-tokens#user-api-tokens)) and this has affected how we do the getStepStartStates (https://github.com/JupiterOne/graph-terraform-cloud/pull/2/files#diff-e3a715a3f673b6ac707d8863cbed5f2e6a41aad3400564d81e9ce3394a39c966R27). The iterateOrganizations part of it is a bit "silly" considering there's only ever going to be a single organization returned, so what we could also do is have the organization name as ENV variable (that would remove the need for calling that endpoint in getStepStartStates). This would however mean that there's an additional step for the customer during the setup, so I'd like to hear others' thoughts on this.

2) We are also ingesting entitlement sets (these are useful for getStepStartStates data) - I'm not sure how useful they're to be stored in the graph, they're ingested currently, but we can easily remove them if necessary. If we want to keep them we'd need to see what'd be the best class for them too.

Graph:
Screen Shot 2021-12-31 at 12 36 08 PM

### CODE REVIEW

File: docs/development.md [hunk 1]
```diff
@@ -1,28 +1,16 @@
...
-
-Supply details about software or tooling (like maybe Docker or Terraform) that
-is needed for development here.
-
-Please supply references to documentation that details how to install those
-dependencies here.
-
-Tools like Node.js and NPM are already covered in the [README](../README.md) so
-don't bother documenting that here.
+This integration focuses on [Terraform Cloud](https://www.terraform.io/cloud) and is using the [Terraform Cloud API](https://www.terraform.io/cloud-docs/api-docs#terraform-cloud-api-documentation) for interacting with the Terraform Cloud resources.
```
Comment (adam-in-ict) [inline docs/development.md:3]: It doesn't look like prettier has run against this file yet. You can manually run it with `yarn format` if the husky Git hooks aren't automatically doing it for you.

**Comment by adam-in-ict** [comment]
A few high level thoughts:
- I'd recommend we drop the "Account" entity, as it doesn't appear to have any data that isn't already captured by the User and Organization entities.  Please correct me if I'm wrong though.
- Regarding the iterateOrganizations and whether or not we should require users to list the organization name instead, I'm in favor of adding the organization name to the .env file and removing the iteration step.  Even if we encourage users to use an organization level token that will only have one organization attached to it, I don't know that we can guarantee that future users won't set this up with a user token, which may have multiple organizations it has access to.  I could see an admin setting this up using their user token and being displeased that both a Dev and Prod organization they have access to got pulled in.  This lets them have the ability to specify the organization they want and saves them from pulling more data than they intend to.

**Review by adam-in-ict — CHANGES_REQUESTED** [review]
I'd recommend we make the following changes before merging:
- Remove Account entity
- Drop iterateOrganizations and add organization name as a .env value.

**Comment by eXtremeX** [comment]
Hey @adam-in-ict, sorry for addressing this late (had quite a few in-progress tasks).

**Drop iterateOrganizations and add organization name as a .env value.**
Awesome, made this change, it made getStepStartStates simpler. Now, just a note, there is some organizations caching involved in code (as seen [here](https://github.com/Creativice-Oy/graph-terraform-cloud/blob/feature/additional-ingestions/src/steps/organization/index.ts#L64) and [here](https://github.com/Creativice-Oy/graph-terraform-cloud/blob/feature/additional-ingestions/src/util/jobState.ts#L10)) - currently I've made [this modification](https://github.com/Creativice-Oy/graph-terraform-cloud/blob/feature/additional-ingestions/src/steps/organization/index.ts#L42) so that we only ever ingest the organization wanted by the user - to prevent the scenario you described above. (We still need to iterate over organizations to pick their IDs unless we also put that in ENV)

This isn't ideal as the caching function expects to work with arrays (more than one organization), I could refactor it to work with just one organization **ever**, but I'm not sure if in the future we may want to accept an array of organization names in the ENV? Might take some further discussions (with more people? or even customers feedback) until we're 100% sure about this.

This works as of now and there are two options (imo):
1. We could accept it as is, only allow customers to input a single organization name and change later if necessary
2. Maybe we stop here, think some more if we'd ever want to accept an array of organization names, and then continue 

**Remove Account entity**
I fully understood your reasons and was about to delete it but then noticed that its data is in fact different and it just seems that the snapshots weren't properly updated. Here's a screenshot. Unless I'm missing something (please let me know if I do), we could keep it up until we encounter something off with it again.

image

**Comment by adam-in-ict** [comment]
@eXtremeX thanks for the updates!

As far as how to proceed with organization, my vote would be for option 1.  If in the future we get users who need to ingest a significant number of organizations, we could change it to accept a comma delimited list of organizations instead of a single, but for now I think that's unnecessary.  If a user needs to ingest just a few organizations, they could always create a new integration instance for the additional organizations.  That might be cleaner anyway for long term use as long as it's a manageable number of organizations.

Regarding the Account I'll have a look at the updated info, but assuming it is distinct with its own data, I agree we should keep it.  Thanks!

File: src/steps/organization/organization.test.ts [hunk 2]
```diff
@@ -145,3 +156,297 @@ describe('#fetchOrganizationMembers', () => {
...
 });
+
+describe('#fetchOrganizationWorkspaces', () => {
+  test('should collect data', async () => {
+    const context = createMockStepExecutionContext<IntegrationConfig>({
+      instanceConfig: integrationConfig,
+    });
+
+    await cacheOrganizationData(context.jobState, [
+      // {
```
Comment (mdaum) [inline src/steps/organization/organization.test.ts:167]: do we need to keep these comments?

File: src/steps/organization/organization.test.ts [hunk 3]
```diff
@@ -145,3 +156,297 @@ describe('#fetchOrganizationMembers', () => {
...
+
+describe('#fetchOrganizationEntitlementSet', () => {
+  test('should collect data', async () => {
+    const context = createMockStepExecutionContext<IntegrationConfig>({
+      instanceConfig: integrationConfig,
+    });
+
+    await cacheOrganizationData(context.jobState, [
+      // {
+      //   organizationName: 'austin-test-org',
```
Comment (mdaum) [inline src/steps/organization/organization.test.ts:291]: do we need these comments?

File: src/steps/organization/organization.test.ts [hunk 4]
```diff
@@ -145,3 +156,297 @@ describe('#fetchOrganizationMembers', () => {
...
+describe('#fetchOrganizationTeams', () => {
+  test('should collect data', async () => {
+    const context = createMockStepExecutionContext<IntegrationConfig>({
+      instanceConfig: integrationConfig,
+    });
+
+    await cacheOrganizationData(context.jobState, [
+      // {
+      //   organizationName: 'austin-test-org',
+      //   organizationExternalId: 'org-Ljx2Ap3vszNhHWWv',
```
Comment (mdaum) [inline src/steps/organization/organization.test.ts:373]: do we need these comments?

File: src/steps/workspace/workspace.test.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+describe('#fetchWorkspaceResources', () => {
+  test('should collect data', async () => {
+    const context = createMockStepExecutionContext<IntegrationConfig>({
+      instanceConfig: integrationConfig,
+    });
+
+    await cacheOrganizationData(context.jobState, [
+      // {
+      //   organizationName: 'austin-test-org',
```
Comment (mdaum) [inline src/steps/workspace/workspace.test.ts:19]: needed?

**Review by mdaum — APPROVED** [review]
looks fine to me. a few places where I think we might want to clean up comments in the tests though

File: docs/jupiterone.md [hunk 6]
```diff
@@ -91,18 +76,28 @@ https://github.com/JupiterOne/sdk/blob/master/docs/integrations/development.md
 
 The following entities are created:
 
-| Resources    | Entity `_type`     | Entity `_class`           |
-| ------------ | ------------------ | ------------------------- |
-| Organization | `tfe_organization` | `Account`, `Organization` |
-| User         | `tfe_user`         | `User`                    |
+| Resources       | Entity `_type`           | Entity `_class`           |
+| --------------- | ------------------------ | ------------------------- |
+| Account         | `tfe_account`            | `User`                    |
```
Comment (adam-in-ict) [inline docs/jupiterone.md:81]: I think the _class should be 'Account' if we're keeping this entity.

File: docs/jupiterone.md [hunk 7]
```diff
@@ -91,18 +76,28 @@ https://github.com/JupiterOne/sdk/blob/master/docs/integrations/development.md
...
+| Entitlement Set | `tfe_entitlement_set`    | `Entity`                  |
+| Organization    | `tfe_organization`       | `Account`, `Organization` |
```
Comment (adam-in-ict) [inline docs/jupiterone.md:83]: Do we want to remove 'Account' from this entity's _class list if we're keeping the tfe_account entity?

File: src/steps/organization/organization.test.ts (hunk 2 above)
Comment (eXtremeX) [reply src/steps/organization/organization.test.ts]: Good catch, they were necessary until the previous change and I forgot to remove them after making sure tests work without them. Thanks for marking all of those!

File: docs/jupiterone.md (hunk 6 above)
Comment (eXtremeX) [reply docs/jupiterone.md]: Agree!

File: docs/jupiterone.md (hunk 7 above)
Comment (eXtremeX) [reply docs/jupiterone.md]: Same here!

**Review by adam-in-ict — APPROVED** [review]

**Pull request merged**